### PR TITLE
[react-native] Add ability to type additional section data

### PR DESCRIPTION
**Pull request by NikolaIliev**
Added a second generic type param to SectionList and all SectionList-related exported types.

The idea is to be able to type out any additional data you may have in your Sections, usually used when rendering the section header/footer etc.

Previously, any additional data would be of type `any` and the user would need to type it manually in callbacks such as `renderSectionHeader`.

The changes are backwards-compatible as I've added a default value for this new generic param. The default value implements the previous behaviour - `{ [key: string]: any }`

Example:

```tsx
import { SectionList, SectionListData } from 'react-native'

type Person = { name: string, age: number }
type SectionT = { title: string }

const sections: SectionListData<Person, SectionT>[] = [
  {
     data: [ ... ],
     title: 'Students'
  },
  {
    data: [...],
    title: 'Professors'
  }
]

// The SectionT generic parameter is inferred by TS
// You are now forced to accept a { section: SectionListData<Person, SectionT> } shape
const renderSectionHeader = ({ section }: { section: SectionListData<Person, SectionT> }) => {
   // -----> section.title is string instead of any <------
   return <Text>{section.title}</Text>
}

<SectionList
  ...
  sections={sections}
  renderSectionHeader={renderSectionHeader}
/>

```

Once again, this is backwards compatible - we may choose not to pass `SectionT` when defining `sections`.

```tsx
import { SectionList, SectionListData } from 'react-native'

type Person = { name: string, age: number }
// type SectionT = { title: string }

const sections: SectionListData<Person, /* SectionT */>[] = [
  {
     data: [ ... ],
     title: 'Students'
  },
  {
    data: [...],
    title: 'Professors'
  }
]

const renderSectionHeader = ({ section }: { section: SectionListData<Person> }) => {
   // -----> section.title is any <------
   return <Text>{section.title}</Text>
}

<SectionList
  ...
  sections={sections}
  renderSectionHeader={renderSectionHeader}
/>

```

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://reactnative.dev/docs/sectionlist#section
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)

### CODE REVIEW

File: types/react-native/test/index.tsx [hunk 1]
```diff
@@ -581,6 +582,69 @@ export class SectionListTest extends React.Component<SectionListProps<string>, {
...
+
+        return (
+            <React.Fragment>
+                <Button title="Press" onPress={this.scrollMe} />
+
+                <SectionList
+                    ref={this.myList}
+                    sections={sections}
+                    renderSectionHeader={({ section }) =>
+                        section.displayTitle ? (
```
Comment (alloy) [inline types/react-native/test/index.tsx:629]: Can you add an assertion here to test the type of `section`?

File: types/react-native/test/index.tsx [hunk 2]
```diff
@@ -581,6 +582,69 @@ export class SectionListTest extends React.Component<SectionListProps<string>, {
...
+                            <View>
+                                <Text>{section.title}</Text>
+                            </View>
+                        ) : null
+                    }
+                    renderItem={(info: SectionListRenderItemInfo<string>) => (
```
Comment (alloy) [inline types/react-native/test/index.tsx:635]: Similarly, can you remove the explicit typing of the parameter and add an assertion here to test the type of `info`?

File: types/react-native/index.d.ts [hunk 3]
```diff
@@ -4535,7 +4539,8 @@ export interface SectionListScrollParams {
     viewPosition?: number;
 }
 
-export class SectionList<SectionT = any> extends React.Component<SectionListProps<SectionT>> {
+export class SectionList<ItemT = any, SectionT = DefaultSectionT> extends React.Component<SectionListProps<ItemT,
```
Comment (alloy) [inline types/react-native/index.d.ts:4542]: This does make it a breaking change, if somebody _had_ specified a type here, no? Having said that, I’m fine with making the change.

**Review by alloy — CHANGES_REQUESTED** [review]
This looks great! Only question is to improve the tests slightly by asserting on what TS gives us as callback parameter types.

File: types/react-native/index.d.ts (hunk 3 above)
Comment (NikolaIliev) [reply types/react-native/index.d.ts:4542]: I'm not sure. As far as I can tell, this generic param was called SectionT but in fact, it is used as ItemT. So, essentially, we're adding a new, second param and simply renaming the first while retaining its semantics and usage.

Comment (alloy) [reply types/react-native/index.d.ts:4542]: Ah gotcha 👍

**Review by alloy — APPROVED** [review]
Sweet, thanks! 🙏

**Pull request merged**